reconcile/utils/gitlab_api.py

```diff
@@ -100,6 +100,22 @@ def __init__(
             gitlab_request.labels(integration=INTEGRATION_NAME).inc()
             self.project = self.gl.projects.get(project_id)
         self.saas_files = saas_files
+        self._main_branch = None
+
+    @property
+    def main_branch(self) -> str:
+        if self._main_branch is None:
+            self._main_branch = self._determine_main_branch()
+        return self._main_branch or "master"
+
+    def _determine_main_branch(self) -> Optional[str]:
+        if self.project:
+            main_branch_obj = next(
+                (b for b in self.project.branches.list() if b.default), None
+            )
+            if main_branch_obj:
+                return main_branch_obj.name
+        return None
```

Comment:
It seems we want to cache `main_branch` computation, right?
How about using `cached_property`?

```python
    DEFAULT_MAIN_BRANCH = "master"

    @cached_property
    def main_branch(self) -> str:
        if self.project:
            return next(
                (b.name for b in self.project.branches.list() if b.default),
                DEFAULT_MAIN_BRANCH,
            )
        return DEFAULT_MAIN_BRANCH
```

Reply:
there are some code paths where the projects becomes available only later. so i want to avoid caching a default branch name when there might be a better one on the project

Reply:
then how about this:

```python
    @cached_property
    def project_main_branch(self) -> str:
        return next(
            (b.name for b in self.project.branches.list() if b.default),
            DEFAULT_MAIN_BRANCH,
        )

    @property
    def main_branch(self) -> str:
        return this.project_main_branch if self.project else DEFAULT_MAIN_BRANCH
```

Reply:
i like this, i will use it.

in general i think the gitlab_api.py code needs some refactoring, because it mixes pure API things with projects things. being able to switch out the project behind the scene makes me kind of nervous.

in the meantime, your code will work nicely. thank you